File: pwa/src/maps/mindMapService.js
import {
  applyMapMutation,
  buildMapSummary,
  compareMapSummariesByRecentUpdate,
  cloneMapDocument,
  collectTaskEntries,
  createMapDocument,
} from './model.js';

export class MindMapService {
  constructor(repository) {
    this.repository = repository;
    this.snapshotsByPath = new Map();
  }

  async listMaps(forceRefresh = false) {
    const listed = await this.repository.listFiles();
    if (!listed.ok) {
      return listed;
    }

    const loadedSnapshots = [];
    const unreadableMaps = [];
    for (const file of listed.value) {
      const loaded = await this.loadMap(file.filePath, forceRefresh);
      if (!loaded.ok) {
        if (loaded.error?.code === 'UNREADABLE_MAP') {
          this.snapshotsByPath.delete(file.filePath);
          unreadableMaps.push(buildUnreadableMapEntry(loaded.error));
          continue;
        }

        return loaded;
      }

      loadedSnapshots.push(loaded.value);
    }

    return {
      ok: true,
      value: {
        snapshots: loadedSnapshots,
        unreadableMaps: unreadableMaps.sort(compareUnreadableMapEntries),
      },
    };
  }

  async loadMap(filePath, forceRefresh = false) {
    if (!forceRefresh && this.snapshotsByPath.has(filePath)) {
      return {
        ok: true,
        value: this.snapshotsByPath.get(filePath),
      };
    }

    const loaded = await this.repository.loadMap(filePath);
    if (!loaded.ok) {
      if (loaded.error?.code === 'UNREADABLE_MAP') {
        this.snapshotsByPath.delete(filePath);
      }
      return loaded;
    }

    this.snapshotsByPath.set(filePath, loaded.value);
    return loaded;
  }

  getCachedSnapshot(filePath) {
    return this.snapshotsByPath.get(filePath) ?? null;
  }

  hydrateSnapshots(snapshots) {
    this.snapshotsByPath.clear();
    snapshots.forEach((snapshot) => {
      if (snapshot?.filePath) {
        this.snapshotsByPath.set(snapshot.filePath, snapshot);
      }
    });
  }

  buildSummaries(snapshots) {
    return snapshots
      .map((snapshot) => buildMapSummary(snapshot))
      .sort(compareMapSummariesByRecentUpdate);
  }

  buildTaskEntries(filter = 'open', snapshots = Array.from(this.snapshotsByPath.values())) {
    return snapshots.flatMap((snapshot) => collectTaskEntries(snapshot, filter));
  }

  async createMap(filePath, mapName, commitMessage) {
    const document = createMapDocument(mapName);
    const saved = await this.repository.createMap(filePath, document, commitMessage);
    if (!saved.ok) {
      return saved;
    }

    const fileName = filePath.split('/').pop() || filePath;
    const snapshot = {
      filePath,
      fileName,
      mapName,
      document,
      revision: saved.revision,
      loadedAt: Date.now(),
    };
    this.snapshotsByPath.set(filePath, snapshot);
    return {
      ok: true,
      value: snapshot,
    };
  }

  async loadAttachment(mapFilePath, relativePath, mediaType) {
    return this.repository.loadAttachment(mapFilePath, relativePath, mediaType);
  }

  async deleteMap(filePath, commitMessage) {
    const snapshot = this.snapshotsByPath.get(filePath);
    if (!snapshot) {
      return {
        ok: false,
        error: {
          code: 'NOT_FOUND',
          message: `Map "${filePath}" is not loaded.`,
          retriable: false,
        },
      };
    }

    const deleted = await this.repository.deleteMap(filePath, snapshot.revision, commitMessage);
    if (!deleted.ok) {
      return deleted;
    }

    this.snapshotsByPath.delete(filePath);
    return {
      ok: true,
    };
  }

  async saveResolved(filePath, document, revision, commitMessage) {
    const saved = await this.repository.saveMap(filePath, document, revision, commitMessage);
    if (!saved.ok) {
      return saved;
    }

    const existing = this.snapshotsByPath.get(filePath);
    const fileName = existing?.fileName || filePath.split('/').pop() || filePath;
    const mapName = existing?.mapName || fileName.replace(/\.json$/i, '');

    const snapshot = {
      filePath,
      fileName,
      mapName,
      document,
      revision: saved.revision,
      loadedAt: Date.now(),
    };
    this.snapshotsByPath.set(filePath, snapshot);
    return {
      ok: true,
      value: snapshot,
    };
  }

  async applyMutation(filePath, mutation, commitMessage) {
    const latest = await this.loadMap(filePath, false);
    if (!latest.ok) {
      return latest;
    }

    const attemptedDocument = cloneMapDocument(latest.value.document);
    const applied = applyMapMutation(attemptedDocument, mutation);
    if (!applied.ok) {
      return {
        ok: false,
        error: {
          ...applied.error,
          meta: {
            filePath,
            mutation,
            commitMessage,
          },
        },
      };
    }

    const firstSave = await this.repository.saveMap(
      filePath,
      attemptedDocument,
      latest.value.revision,
      commitMessage,
    );
    if (firstSave.ok) {
      const snapshot = {
        ...latest.value,
        document: attemptedDocument,
        revision: firstSave.revision,
        loadedAt: Date.now(),
      };
      this.snapshotsByPath.set(filePath, snapshot);
      return {
        ok: true,
        value: {
          snapshot,
          mutation: applied.value,
        },
      };
    }

    if (firstSave.error.code !== 'STALE_STATE') {
      return firstSave;
    }

    const refreshed = await this.loadMap(filePath, true);
    if (!refreshed.ok) {
      return refreshed;
    }

    const retriedDocument = cloneMapDocument(refreshed.value.document);
    const retriedApply = applyMapMutation(retriedDocument, mutation);
    if (!retriedApply.ok) {
      return {
        ok: false,
        error: {
          ...retriedApply.error,
          meta: {
            filePath,
            mutation,
            commitMessage,
          },
        },
      };
    }

    const retrySave = await this.repository.saveMap(
      filePath,
      retriedDocument,
      refreshed.value.revision,
      commitMessage,
    );
    if (!retrySave.ok) {
      return {
        ok: false,
        error: {
          code: 'CONFLICT_UNRESOLVED',
          message: `Map "${filePath}" conflicted twice and needs manual resolution.`,
          retriable: false,
          cause: retrySave.error,
          meta: {
            filePath,
            mutation,
            commitMessage,
          },
        },
      };
    }

    const snapshot = {
      ...refreshed.value,
      document: retriedDocument,
      revision: retrySave.revision,
      loadedAt: Date.now(),
    };
    this.snapshotsByPath.set(filePath, snapshot);
    return {
      ok: true,
      value: {
        snapshot,
        mutation: retriedApply.value,
      },
    };
  }
}

function buildUnreadableMapEntry(error) {
  return {
    filePath: error.filePath,
    fileName: error.fileName,
    mapName: error.mapName,
    revision: error.revision,
    reason: error.reason,
    message: error.message,
    rawText: error.rawText,
  };
}

function compareUnreadableMapEntries(left, right) {
  return String(left?.fileName ?? '').localeCompare(String(right?.fileName ?? ''));
}

File: android/app/src/main/java/com/systemssanity/focus/domain/sync/MindMapService.kt
package com.systemssanity.focus.domain.sync

import com.systemssanity.focus.data.github.UnreadableMapException
import com.systemssanity.focus.domain.maps.CommitMessages
import com.systemssanity.focus.domain.maps.DeletedAttachmentRef
import com.systemssanity.focus.domain.maps.MapMutation
import com.systemssanity.focus.domain.maps.MapMutationEngine
import com.systemssanity.focus.domain.maps.MapMutationRejectedException
import com.systemssanity.focus.domain.maps.MutationApplyResult
import com.systemssanity.focus.domain.maps.UnreadableMapEntry
import com.systemssanity.focus.domain.maps.collectDeletedAttachmentRefs
import com.systemssanity.focus.domain.maps.deletedAttachmentKey
import com.systemssanity.focus.domain.maps.normalizeDeletedAttachmentRefs
import com.systemssanity.focus.domain.model.MindMapJson
import com.systemssanity.focus.domain.model.MapSnapshot
import com.systemssanity.focus.domain.model.MindMapDocument
import java.util.UUID

class MindMapService(private val repository: MindMapRepositoryGateway) {
    private val snapshotsByPath = LinkedHashMap<String, MapSnapshot>()

    suspend fun listMaps(forceRefresh: Boolean = false): Result<MindMapListResult> =
        repository.listFiles().mapCatching { files ->
            val snapshots = mutableListOf<MapSnapshot>()
            val unreadableMaps = mutableListOf<UnreadableMapEntry>()
            files.forEach { (_, filePath) ->
                loadMap(filePath, forceRefresh)
                    .onSuccess { snapshots += it }
                    .onFailure { error ->
                        if (error is UnreadableMapException) {
                            snapshotsByPath.remove(filePath)
                            unreadableMaps += error.toUnreadableMapEntry()
                        } else {
                            throw error
                        }
                    }
            }
            MindMapListResult(
                snapshots = snapshots,
                unreadableMaps = unreadableMaps.sortedBy { it.fileName.lowercase() },
            )
        }

    suspend fun loadMap(filePath: String, forceRefresh: Boolean = false): Result<MapSnapshot> {
        if (!forceRefresh) {
            snapshotsByPath[filePath]?.let { return Result.success(it) }
        }
        return repository.loadMap(filePath)
            .onSuccess { snapshotsByPath[filePath] = it }
            .onFailure { error ->
                if (error is UnreadableMapException) snapshotsByPath.remove(filePath)
            }
    }

    suspend fun createMap(filePath: String, mapName: String, commitMessage: String): Result<MapSnapshot> {
        val document = MindMapJson.normalize(MindMapDocument(
            rootNode = com.systemssanity.focus.domain.model.Node(
                uniqueIdentifier = UUID.randomUUID().toString(),
                name = mapName,
            ),
        ))
        return repository.createMap(filePath, document, commitMessage).map { revision ->
            MapSnapshot(
                filePath = filePath,
                fileName = filePath.substringAfterLast('/'),
                mapName = mapName,
                document = document,
                revision = revision,
                loadedAtMillis = System.currentTimeMillis(),
            ).also { snapshotsByPath[filePath] = it }
        }
    }

    suspend fun deleteMap(filePath: String, commitMessage: String): Result<Unit> =
        runCatching {
            val latest = loadMap(filePath, forceRefresh = true).getOrThrow()
            deleteAttachmentRefs(
                refs = collectDeletedAttachmentRefs(latest.document.rootNode),
                commitMessage = commitMessage,
            )
            repository.deleteMap(filePath, latest.revision, commitMessage).getOrThrow()
            snapshotsByPath.remove(filePath)
            Unit
        }

    suspend fun renameMap(
        oldFilePath: String,
        newFilePath: String,
        document: MindMapDocument,
        oldRevision: String,
        commitMessage: String,
    ): Result<MapSnapshot> =
        repository.renameMap(oldFilePath, newFilePath, MindMapJson.normalize(document), oldRevision, commitMessage)
            .map { revision ->
                val fileName = newFilePath.substringAfterLast('/')
                MapSnapshot(
                    filePath = newFilePath,
                    fileName = fileName,
                    mapName = fileName.removeSuffix(".json"),
                    document = MindMapJson.normalize(document),
                    revision = revision,
                    loadedAtMillis = System.currentTimeMillis(),
                ).also {
                    snapshotsByPath.remove(oldFilePath)
                    snapshotsByPath[newFilePath] = it
                }
            }

    suspend fun loadAttachment(nodeId: String, relativePath: String, mediaType: String) =
        repository.loadAttachment(nodeId, relativePath, mediaType)

    suspend fun uploadAttachment(nodeId: String, relativePath: String, base64Content: String, commitMessage: String) =
        repository.uploadAttachment(nodeId, relativePath, base64Content, commitMessage)

    suspend fun deleteAttachment(nodeId: String, relativePath: String, expectedRevision: String?, commitMessage: String) =
        repository.deleteAttachment(nodeId, relativePath, expectedRevision, commitMessage)

    suspend fun applyMutation(filePath: String, mutation: MapMutation): Result<MapSnapshot> =
        runCatching {
            val latest = loadMap(filePath).getOrThrow()
            val applied = MapMutationEngine.apply(latest.document, mutation)
            if (applied is MutationApplyResult.Rejected) {
                throw MapMutationRejectedException(applied.code, applied.message)
            }
            check(applied is MutationApplyResult.Applied)

            val deletedAttachmentKeys = LinkedHashSet<String>()
            deleteAttachmentRefs(applied.result.deletedAttachments, mutation.commitMessage, deletedAttachmentKeys)
            val firstSave = repository.saveMap(
                filePath = filePath,
                document = applied.result.document,
                revision = latest.revision,
                commitMessage = mutation.commitMessage,
            )
            var savedDocument = applied.result.document
            val revision = firstSave.getOrElse { firstError ->
                if (!repository.run { firstError.isStaleState() }) throw firstError
                val refreshed = loadMap(filePath, forceRefresh = true).getOrThrow()
                val retried = MapMutationEngine.apply(refreshed.document, mutation)
                if (retried is MutationApplyResult.Rejected) {
                    throw MapMutationRejectedException(retried.code, retried.message)
                }
                check(retried is MutationApplyResult.Applied)
                savedDocument = retried.result.document
                deleteAttachmentRefs(retried.result.deletedAttachments, mutation.commitMessage, deletedAttachmentKeys)
                repository.saveMap(filePath, retried.result.document, refreshed.revision, mutation.commitMessage).getOrThrow()
            }

            latest.copy(
                document = savedDocument,
                revision = revision,
                loadedAtMillis = System.currentTimeMillis(),
            ).also { snapshotsByPath[filePath] = it }
        }

    suspend fun saveResolved(
        filePath: String,
        mapName: String,
        document: MindMapDocument,
        revision: String,
    ): Result<MapSnapshot> =
        repository.saveMap(filePath, MindMapJson.normalize(document), revision, CommitMessages.conflictResolve(mapName))
            .map { savedRevision ->
                val fileName = filePath.substringAfterLast('/')
                MapSnapshot(
                    filePath = filePath,
                    fileName = fileName,
                    mapName = mapName.ifBlank { fileName.removeSuffix(".json") },
                    document = MindMapJson.normalize(document),
                    revision = savedRevision,
                    loadedAtMillis = System.currentTimeMillis(),
                ).also { snapshotsByPath[filePath] = it }
            }

    fun hydrateSnapshots(snapshots: List<MapSnapshot>) {
        snapshotsByPath.clear()
        snapshots.forEach { snapshotsByPath[it.filePath] = it }
    }

    fun replaceCachedSnapshot(snapshot: MapSnapshot) {
        snapshotsByPath[snapshot.filePath] = snapshot
    }

    fun removeCachedSnapshot(filePath: String) {
        snapshotsByPath.remove(filePath)
    }

    fun cachedSnapshots(): List<MapSnapshot> = snapshotsByPath.values.toList()

    fun isNotFound(error: Throwable): Boolean =
        repository.run { error.isNotFound() }

    fun isStaleState(error: Throwable): Boolean =
        repository.run { error.isStaleState() }

    private suspend fun deleteAttachmentRefs(
        refs: List<DeletedAttachmentRef>,
        commitMessage: String,
        deletedKeys: MutableSet<String> = LinkedHashSet(),
    ) {
        normalizeDeletedAttachmentRefs(refs).forEach { ref ->
            if (deletedKeys.add(deletedAttachmentKey(ref))) {
                repository.deleteAttachment(
                    nodeId = ref.nodeId,
                    relativePath = ref.relativePath,
                    expectedRevision = null,
                    commitMessage = commitMessage,
                ).getOrThrow()
            }
        }
    }
}

data class MindMapListResult(
    val snapshots: List<MapSnapshot>,
    val unreadableMaps: List<UnreadableMapEntry> = emptyList(),
)

internal fun UnreadableMapException.toUnreadableMapEntry(): UnreadableMapEntry =
    UnreadableMapEntry(
        filePath = filePath,
        fileName = fileName,
        mapName = mapName,
        revision = revision,
        reason = reason,
        message = message.orEmpty(),
        rawText = rawText,
    )
